src/main/java/com/leverx/learningmanagementsystem/service/impl/StudentServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+
+    @Override
+    public Student getById(UUID id) {
+        log.info("Get student by id: {}", id);
+        return studentRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Student with id " + id + " not found"));
+    }
+
+    @Override
+    public List<Student> getAll() {
+        log.info("Get all students");
+        return studentRepository.findAll();
+    }
+
+    @Override
+    @Transactional
+    public Student create(CreateStudentDto createStudentDto) {
+
+        Student student = studentMapper.toStudent(createStudentDto);
+        log.info("Create student: {}", student);
+        saveStudent(student, createStudentDto);
+
+        return student;
+    }
+
+    @Override
+    @Transactional
+    public Student update(UUID id, CreateStudentDto updateStudentDto) {
+
+        if (studentRepository.findById(id).isEmpty()) {
+            throw new EntityNotFoundException("Student with id " + id + " not found");
+        }
+
+        Student student = studentMapper.toStudent(updateStudentDto);
+        student.setId(id);
+
+        log.info("Update student: {}", student);
+        saveStudent(student, updateStudentDto);
+
+        return student;
+    }
+
+    @Transactional
+    public void enrollForCourse(UUID studentId, UUID courseId) {
+        Student student = getById(studentId);
+        Course course = courseService.getById(courseId);
+
+        if (student.getCourses().contains(course)) {
+            throw new StudentAlreadyEnrolledException("Student with id " + studentId
+                    + " already enrolled for course with id = " + courseId);
+        }
+
+        if (student.getCoins().compareTo(course.getPrice()) < 0) {
+            throw new NotEnoughCoinsException("Student with id = " + studentId
+                    + " doesn't have enough coins to enroll for course with id = " + courseId);
+        }
+
+        student.setCoins(student.getCoins().subtract(course.getPrice()));
+        student.getCourses().add(course);
+        course.setCoinsPaid(course.getCoinsPaid().add(course.getPrice()));
+        course.getStudents().add(student);
+
+        studentRepository.save(student);
+        courseRepository.save(course);
+    }
+
+    @Override
+    public void delete(UUID id) {
+        getById(id);
+        log.info("Delete student: {}", id);
+        studentRepository.deleteById(id);
+    }
+
+    private void saveStudent(Student student, CreateStudentDto createStudentDto) {
+
+        log.info("Fetching courses for student with id = {}", student.getId());
+        Set<Course> courses = new HashSet<>(courseRepository.findAllById(createStudentDto.courseIds()));
+
+        if (createStudentDto.courseIds() != null && courses.size() != createStudentDto.courseIds().size()) {
+            throw new IncorrectResultSizeException("Some of requested courses don't exist");
+        }
+
+        student.setCourses(courses);
+        studentRepository.save(student);
+    }
```

Comment:
It's better to create `StudentValidator` class which will contain logic for validation on create and update requests